Core/src/Utilities/FpeMonitor.cpp

```diff
@@ -10,81 +10,255 @@
+  Result result;
+  for (unsigned int i = 0; i < m_counts.size(); i++) {
+    result.m_counts[i] = m_counts[i] + with.m_counts[i];
   }
 
-  if (check(FE_DIVBYZERO)) {
-    std::cout << "FE_DIVBYZERO" << std::endl;
-  }
+  std::copy(m_stracktraces.begin(), m_stracktraces.end(),
+            std::back_inserter(result.m_stracktraces));
+  std::copy(with.m_stracktraces.begin(), with.m_stracktraces.end(),
+            std::back_inserter(result.m_stracktraces));
 
-  if (check(FE_INVALID)) {
-    std::cout << "FE_INVALID" << std::endl;
-  }
+  // @TODO: Optimize this by not merging first and then deduplicating
+  result.deduplicate();
+
+  // @TODO: Merge stack traces
+  return result;
+}
+
+const FpeMonitor::Result &FpeMonitor::result() const {
+  return m_result;
+}
+
+FpeMonitor::Result &FpeMonitor::result() {
+  return m_result;
+}
+
+unsigned int FpeMonitor::Result::count(FpeType type) const {
+  return m_counts.at(static_cast<uint32_t>(type));
+}
+
+unsigned int FpeMonitor::Result::numStackTraces() const {
+  return m_stracktraces.size();
+}
 
-  if (check(FE_UNDERFLOW)) {
-    std::cout << "FE_UNDERFLOW" << std::endl;
+const std::vector<FpeMonitor::Result::FpeInfo>
+    &FpeMonitor::Result::stackTraces() const {
+  return m_stracktraces;
+}
+
+bool FpeMonitor::Result::encountered(FpeType type) const {
+  return count(type) > 0;
+}
+
+void FpeMonitor::Result::summary(std::ostream &os) const {
+  os << "FPE result summary:\n";
+  static const std::vector<FpeType> types = {
+      FpeType::INTDIV, FpeType::INTOVF, FpeType::FLTDIV, FpeType::FLTOVF,
+      FpeType::FLTUND, FpeType::FLTRES, FpeType::FLTINV, FpeType::FLTSUB};
+
+  for (auto type : types) {
+    os << "- " << type << ": " << count(type) << "\n";
   }
 
-  if (check(FE_INEXACT)) {
-    std::cout << "FE_INEXACT" << std::endl;
+  os << "\nStack traces:\n";
+  for (const auto &[count, type, st] : stackTraces()) {
+    os << "- " << type << ": (" << count << " times)\n" << st;
   }
+}
 
-  std::cout << boost::stacktrace::stacktrace();
+void FpeMonitor::Result::deduplicate() {
+  std::vector<FpeInfo> copy = std::move(m_stracktraces);
+  m_stracktraces.clear();
 
-  std::abort();
+  for (auto &info : copy) {
+    // auto type = it.first;
+    // auto &st = it.second;
+    auto it = std::find_if(m_stracktraces.begin(), m_stracktraces.end(),
+                           [&info](const FpeInfo &el) {
+                             return el.type == info.type && el.st == info.st;
+                           });
+    if (it != m_stracktraces.end()) {
+      it->count += info.count;
+      continue;
+    }
+    m_stracktraces.push_back({info.count, info.type, std::move(info.st)});
+  }
 }
 
-}  // namespace
-
 FpeMonitor::FpeMonitor()
-    : FpeMonitor{FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW} {}
+    : FpeMonitor{FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW | FE_UNDERFLOW} {}
 
 FpeMonitor::FpeMonitor(int excepts) : m_excepts(excepts) {
-  enable(m_excepts);
+  enable();
 }
 
 FpeMonitor::~FpeMonitor() {
```

Comment:
should the nested monitor communicate its counts to the next element in the stack?

Reply:
I've been thinking about his. I chose not to do this automatically, but have a way to merge the results explicitly.